refactor(extractor/babel)!: sealed, non-fragile babel extractor options (#1367)
lingui/js-lingui: website/docs/guides/advanced-custom-extractor.md

> # Advanced: Custom Extractor

Reviewer: ```suggestion
# Custom Extractor
```

I'd suggest removing this prefix since such a heading format is something unusual for this site

\+ rename file

Author: That's exactly how i wanted to write this. There will be more in the future. If you want you can create a separate group in the sidebar for that. But this is totally fine to have "advanced" https://nextjs.org/docs/advanced-features/compiler

Reviewer: Let's create a separate "Advanced" category. This could be a sub-category inside the "Guides" category.

If you not project not playing well with Lingui's Extractor you can write your custom extractor implementation.

> That's might be the case if you use some experimental features (stage0 - stage2) or frameworks with custom syntax such as Vue or Svelte.

Reviewer: ```suggestion
That might be the case if you use some experimental features (stage0 - stage2) or frameworks with custom syntax such as Vue.js or Svelte.
```

```ts title="./my-custom-extractor.ts"
import babel from "@lingui/cli/api/extractors/babel"

export const extractor: ExtractorType = {
  match(filename: string) {
      return filename.endsWith(".custom");
  },
  extract(filename: string, code: string, onMessageExtracted) {
    // transform to plain JS + Sourcemaps
     const {code, sourcemaps} = transformMyCustomFileToJs(filename, code);

     // reuse extractor from cli
     return babel(filename, code, onMessageExtracted, {sourcemaps})
  }
}
```

Then in your `lingui.config.ts`

```ts title="lingui.config.ts"
import {extractor} from './my-custom-extractor.ts'
module.exports = {
  [...]
  extractors: [extractor]
}
```

:::caution Important

> If you use typescript for creating your extractor, you should also use `ts` extension for your lingui config file.

Reviewer: ```suggestion
If you use TypeScript to create your extractor, you should use the `ts` extension for your Lingui configuration file.
```
